### PR TITLE
Batching in Router and PipelineStorageStream

### PR DESCRIPTION
**Pull request by alicefuzier**
The router is currently sending documents to index in elastic search using the batching API, but it sends then one at a time, so it's effectively not batching. Because messages can only be deleted from the queue when indexing and sending on the id has succeeded, batching affects the whole stream, as we need to hold on to the original message to be able to delete it.

This PR is adding logic to group the messages into batches in a `PipelineStorageStream` which is a decorator for `SQSStream`. In this PR `PipelineStorageStream` is used in the router, but it should be easy to swap usages of SQSStream for `PipelineStorageStream` without having to change the code except for configuration and infrastructure

### CODE REVIEW

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  private def batchAndSendFlow(implicit indexable: Indexable[D]) = {
+    Flow[(Message, Option[D])]
+      .collect { case (message, Some(document)) => Bundle(message, document) }
+      .groupedWithin(
+        config.batchSize,
+        config.flushInterval
+      )
+      .mapAsyncUnordered(10) { msgs =>
+        storeDocuments(msgs.toList)
+      }
```
Comment (alexwlchan) [inline common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:65]: If I've read this correctly, this groups up to 10 messages at once? Is it worth exposing this as a top level config item, so we can tweak it?

File: common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+class PipelineStorageStreamTest
+    extends AnyFunSpec
+    with ElasticIndexerFixtures
+    with IdentifiersGenerators
+    with RandomGenerators
+    with ElasticsearchFixtures
+    with SQS
+    with Akka {
+
+  val pipelineStorageConfig = PipelineStorageConfig(1, 1 seconds, 10)
```
Comment (alexwlchan) [inline common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala:37]: keyword arguments plz

File: common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                            pipelineStorageConfig: PipelineStorageConfig =
+                              pipelineStorageConfig)(
+    testWith: TestWith[
+      (PipelineStorageStream[NotificationMessage, SampleDocument, String],
+       QueuePair,
+       MemoryMessageSender),
+      R]): R =
+    withActorSystem { implicit ac =>
+      withLocalSqsQueuePair(visibilityTimeout) {
+        case q @ QueuePair(queue, _) =>
```
Comment (alexwlchan) [inline common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala:51]: Why are you creating a queue pair if you never use the DLQ?

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala (hunk 1 above)
Comment (alicefuzier) [reply common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:65]: ah yes, good spot thanks

File: common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala (hunk 3 above)
Comment (alicefuzier) [reply common/pipeline_storage/src/test/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStreamTest.scala:51]: we do, just not here in the `with` method. We pass the whole `QueuePair` in `testWith` so each test can use it

**Review by kenoir — APPROVED** [review]

**Pull request merged**